### PR TITLE
planner: add 2pb logic and fix some bugs for partitionTopN

### LINKED ISSUE #42321: Query returns wrong result if `tidb_opt_derive_topn` is on
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
```
mysql> show create table t;
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                        |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `id1` int(11) NOT NULL,
  `id2` int(11) NOT NULL,
  `value1` int(11) DEFAULT NULL,
  `value2` int(11) DEFAULT NULL,
  PRIMARY KEY (`id1`,`id2`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select * from t;
+-----+-----+--------+--------+
| id1 | id2 | value1 | value2 |
+-----+-----+--------+--------+
|   1 |   1 |      1 |      1 |
|   1 |   2 |      1 |      1 |
|   1 |   3 |      1 |      1 |
|   1 |   4 |      1 |      1 |
|   2 |   1 |      1 |      1 |
|   2 |   2 |      1 |      1 |
|   2 |   3 |      1 |      1 |
|   2 |   4 |      1 |      1 |
+-----+-----+--------+--------+
8 rows in set (0.00 sec)

mysql> set tidb_opt_derive_topn=off;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from (select *, row_number() over (partition by id1) as rownumber from t) DT where rownumber <= 3;
+-----+-----+--------+--------+-----------+
| id1 | id2 | value1 | value2 | rownumber |
+-----+-----+--------+--------+-----------+
|   1 |   1 |      1 |      1 |         1 |
|   1 |   2 |      1 |      1 |         2 |
|   1 |   3 |      1 |      1 |         3 |
|   2 |   1 |      1 |      1 |         1 |
|   2 |   2 |      1 |      1 |         2 |
|   2 |   3 |      1 |      1 |         3 |
+-----+-----+--------+--------+-----------+
6 rows in set (0.01 sec)

mysql> set tidb_opt_derive_topn=on;
Query OK, 0 rows affected (0.01 sec)

mysql> select * from (select *, row_number() over (partition by id1) as rownumber from t) DT where rownumber <= 3;
+-----+-----+--------+--------+-----------+
| id1 | id2 | value1 | value2 | rownumber |
+-----+-----+--------+--------+-----------+
|   1 |   1 |      1 |      1 |         1 |
|   1 |   2 |      1 |      1 |         2 |
|   1 |   3 |      1 |      1 |         3 |
+-----+-----+--------+--------+-----------+
3 rows in set (0.00 sec)
```
### 2. What did you expect to see? (Required)
When set `tidb_opt_derive_topn` to `on` or `off`, the query result should be the same
### 3. What did you see instead (Required)
When set `tidb_opt_derive_topn` to `on`, query result is wrong.
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
master@ fa28db109db7055cead9591cba9c6dcbc48ae009

### PR DESCRIPTION
**Pull request by windtalker**
### What problem does this PR solve?

Issue Number: close #42321, ref #39792

Problem Summary:
partitionTopN lack of 2pb logic and have some bugs.
### What is changed and how it works?

This pr merges the changes in https://github.com/pingcap/tidb/pull/41763 and add the bug fix for #42321

1. add in plan_2_pb, add the partition column for both topn and limit
2. call resolve index for both topn and limit
3. generate partition by column when limit is pushed down
4. show partition by columns info in the result of explain when limit operator has partition by columns
### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/explain.go [hunk 1]
```diff
@@ -364,12 +364,14 @@ func (p *PhysicalSort) ExplainInfo() string {
 
 // ExplainInfo implements Plan interface.
 func (p *PhysicalLimit) ExplainInfo() string {
-	var str strings.Builder
-	str.WriteString("offset:")
-	str.WriteString(strconv.FormatUint(p.Offset, 10))
-	str.WriteString(", count:")
-	str.WriteString(strconv.FormatUint(p.Count, 10))
-	return str.String()
+	buffer := bytes.NewBufferString("")
```
Comment (ghazalfamilyusa) [inline planner/core/explain.go:367]: I think it is better to differentiate Limit with partition by calling it WindowLimit and do not include partition by fields.

File: planner/core/explain.go [hunk 2]
```diff
@@ -960,12 +962,14 @@ func (lt *LogicalTopN) ExplainInfo() string {
 
 // ExplainInfo implements Plan interface.
 func (p *LogicalLimit) ExplainInfo() string {
-	var str strings.Builder
-	str.WriteString("offset:")
-	str.WriteString(strconv.FormatUint(p.Offset, 10))
-	str.WriteString(", count:")
-	str.WriteString(strconv.FormatUint(p.Count, 10))
-	return str.String()
+	buffer := bytes.NewBufferString("")
```
Comment (ghazalfamilyusa) [inline planner/core/explain.go:965]: Same as above. I think it is better to differentiate TopN with partition by calling it WindowTopN and do not include partition by fields.

File: planner/core/explain.go (hunk 1 above)
Comment (windtalker) [reply planner/core/explain.go:367]: You mean adding a new `PhysicalWindowLimit` or just display as `WindowLimit` in the explain info? Adding a new `PhysicalWindowLimit` need a lots of changes, and I think we will miss the deadline if we wants to this :(

File: planner/core/logical_plans.go [hunk 3]
```diff
@@ -163,6 +163,7 @@ type LogicalJoin struct {
 	rightPreferJoinType uint
 
 	EqualConditions []*expression.ScalarFunction
+	// NAEQConditions means null aware equal conditions, which is used for null aware semi joins.
```
[the remark this answers is not in the document]
Comment (windtalker) [reply planner/core/logical_plans.go:166]: No, this is the changes from #41763, I can remove it.

**Review by fixdb — APPROVED** [review]
+1

**Review by AilinKid — APPROVED** [review]

**Pull request merged**